File: node-js/sigv4a_sign.js
const crt = require("aws-crt");
const {HttpRequest} = require("aws-crt/dist/native/http");

function sigV4ASign(method, endpoint, config = crt.auth.AwsSigningConfig) {
    const host = new URL(endpoint).host;
    const request = new HttpRequest(method, endpoint);
    request.headers.add('host', host);

    crt.auth.aws_sign_request(request, config);
    return request.headers;
}

function sigV4ASignBasic(method, endpoint, service) {
    const host = new URL(endpoint).host;
    const request = new HttpRequest(method, endpoint);
    request.headers.add('host', host);

    const config = {
        service: service,
        region: "*",
        algorithm: crt.auth.AwsSigningAlgorithm.SigV4Asymmetric,
        signature_type: crt.auth.AwsSignatureType.HttpRequestViaHeaders,
        signed_body_header: crt.auth.AwsSignedBodyHeaderType.XAmzContentSha256,
        provider: crt.auth.AwsCredentialsProvider.newDefault()
    };

    crt.auth.aws_sign_request(request, config);
    return request.headers;
}

module.exports = {
    sigV4ASign,
    sigV4ASignBasic,
}

File: dotnet/SigV4A/SigV4ASign.cs
using Aws.Crt.Auth;
using Aws.Crt.Http;
using SigV4A.Contracts;

namespace SigV4A;

public class SigV4ASign : ISigV4ASign
{
    public IEnumerable<HttpHeader> SignRequest(string method, string endpoint, string service, Credentials credentials)
    {
        var endpointUri = new Uri(endpoint);
        var request = new HttpRequest
        {
            Method = method,
            Uri = endpointUri.PathAndQuery,
            Headers = new [] { new HttpHeader("host", endpointUri.Host) }
        };

        var config = new AwsSigningConfig
        {
            Service = service,
            Region = "*",
            Algorithm = AwsSigningAlgorithm.SIGV4A,
            SignatureType = AwsSignatureType.HTTP_REQUEST_VIA_HEADERS,
            SignedBodyHeader = AwsSignedBodyHeaderType.X_AMZ_CONTENT_SHA256,
            Credentials = credentials,
        };

        var result = AwsSigner.SignHttpRequest(request, config);
        AwsSigner.CrtSigningResult signingResult = result.Get();
        HttpRequest signedRequest = signingResult.SignedRequest;

        return signedRequest.Headers;
    }
 }